File: agentic-service/app/schemas/pricing_data.py
from typing import Any

from pydantic import BaseModel, ConfigDict, Field, model_validator


class TerrainMultiplier(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    flat: float = 1.0
    hillside: float = 1.25
    coastal: float = 1.35

    @model_validator(mode="before")
    @classmethod
    def _normalize_keys(cls, data: Any) -> Any:
        if isinstance(data, dict):
            return {str(k).lower(): v for k, v in data.items()}
        return data


class PricingItem(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    id: int
    item_name: str = Field(..., alias="itemName")
    category: str
    display_group: str | None = Field(None, alias="displayGroup")
    unit_cost_lkr: float = Field(..., alias="unitCostLkr")
    unit: str
    terrain_multiplier: TerrainMultiplier = Field(default_factory=TerrainMultiplier, alias="terrainMultiplier")
    region: str = "Sri Lanka"
    quality_level: str = Field("Standard", alias="qualityLevel")
    is_active: bool = Field(True, alias="isActive")
    provider: str | None = None
    source_reference: str | None = Field(None, alias="sourceReference")
    updated_at: str | None = Field(None, alias="updatedAt")

    @model_validator(mode="before")
    @classmethod
    def _normalize_keys(cls, data: Any) -> Any:
        if isinstance(data, dict):
            normalized = {}
            key_map = {
                "id": "id",
                "itemname": "item_name",
                "item_name": "item_name",
                "category": "category",
                "displaygroup": "display_group",
                "display_group": "display_group",
                "unitcostlkr": "unit_cost_lkr",
                "unit_cost_lkr": "unit_cost_lkr",
                "unit": "unit",
                "terrainmultiplier": "terrain_multiplier",
                "terrain_multiplier": "terrain_multiplier",
                "region": "region",
                "qualitylevel": "quality_level",
                "quality_level": "quality_level",
                "isactive": "is_active",
                "is_active": "is_active",
                "provider": "provider",
                "sourcereference": "source_reference",
                "source_reference": "source_reference",
                "updatedat": "updated_at",
                "updated_at": "updated_at",
            }
            for k, v in data.items():
                k_clean = str(k).replace("_", "").lower()
                target_key = key_map.get(str(k).lower(), key_map.get(k_clean, k))
                normalized[target_key] = v
            return normalized
        return data

File: HousePlanner.API/Entities/PricingData.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using System.Text.Json.Serialization;

namespace HousePlanner.API.Entities
{
    [Table("PricingData")]
    public class PricingData
    {
        [Key]
        [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
        public int Id { get; set; }

        [Required]
        [StringLength(255)]
        public string ItemName { get; set; } = null!;

        [Required]
        [StringLength(100)]
        public string Category { get; set; } = null!; // e.g. material, labour

        [Required]
        [Column(TypeName = "decimal(18,2)")]
        public decimal UnitCostLkr { get; set; }

        [Required]
        [StringLength(50)]
        public string Unit { get; set; } = null!;

        [Required]
        public TerrainMultiplierData TerrainMultiplier { get; set; } = new TerrainMultiplierData();

        [StringLength(100)]
        public string? DisplayGroup { get; set; }

        [StringLength(100)]
        public string? Provider { get; set; }

        [StringLength(255)]
        public string? ExternalItemId { get; set; }

        [StringLength(255)]
        public string? ExternalItemName { get; set; }

        [StringLength(50)]
        public string? OriginalUnit { get; set; }

        [Column(TypeName = "decimal(18,4)")]
        public decimal? OriginalPrice { get; set; }

        [StringLength(10)]
        public string? OriginalCurrency { get; set; }

        [StringLength(150)]
        public string Region { get; set; } = "Sri Lanka";

        [Required]
        [StringLength(20)]
        public string QualityLevel { get; set; } = "Standard";

        public bool IsActive { get; set; } = true;

        public DateTimeOffset CreatedAt { get; set; } = DateTimeOffset.UtcNow;

        [StringLength(100)]
        public string? UpdatedByUserId { get; set; }

        public DateTimeOffset? ObservedAt { get; set; }

        public DateTimeOffset? EffectiveAt { get; set; }

        [StringLength(2048)]
        public string? SourceUrl { get; set; }

        [StringLength(500)]
        public string? SourceReference { get; set; }

        public DateTimeOffset? ImportedAt { get; set; }

        public DateTimeOffset UpdatedAt { get; set; }

        public ICollection<PricingHistory> History { get; set; } = new List<PricingHistory>();
    }

    public class TerrainMultiplierData
    {
        [JsonPropertyName("flat")]
        public decimal Flat { get; set; } = 1.0m;

        [JsonPropertyName("hillside")]
        public decimal Hillside { get; set; } = 1.25m;

        [JsonPropertyName("coastal")]
        public decimal Coastal { get; set; } = 1.35m;
    }
}
